ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+    if api_key_attrs and hasattr(instance, str(api_key_attrs[0])):
+        api_key = getattr(instance, api_key_attrs[0], None)
+        if api_key:
+            return "...%s" % api_key[-4:]
+    return ""
+
+
+def _tag_openai_token_usage(span, llm_output, propagated_cost=0, propagate=False):
+    """
+    Extract token usage from llm_output, tag on span.
+    Calculate the total cost for each LLM/chat_model, then propagate those values up the trace so that
+    the root span will store the total token_usage/cost of all of its descendants.
+    """
+    for token_type in ("prompt", "completion", "total"):
+        current_metric_value = span.get_metric("langchain.tokens.%s_tokens" % token_type) or 0
+        metric_value = llm_output["token_usage"].get("%s_tokens" % token_type, 0)
+        span.set_metric("langchain.tokens.%s_tokens" % token_type, current_metric_value + metric_value)
+    total_cost = span.get_metric("langchain.tokens.total_cost") or 0
+    if not propagate:
+        try:
+            completion_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"),
+                span.get_metric("langchain.tokens.completion_tokens"),
+                is_completion=True,
+            )
+            prompt_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"), span.get_metric("langchain.tokens.prompt_tokens")
+            )
+            total_cost = completion_cost + prompt_cost
+        except ValueError:
+            # If not in langchain's openai model catalog, assume 0 total cost.
+            pass
+    span.set_metric("langchain.tokens.total_cost", propagated_cost + total_cost)
+    if span._parent is not None:
+        _tag_openai_token_usage(span._parent, llm_output, propagated_cost=propagated_cost + total_cost, propagate=True)
+
+
+@with_traced_module
+def traced_llm_generate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type
+    prompts = args[0]
+    integration = langchain._datadog_integration
+    span = integration.trace(pin, "%s.%s" % (instance.__module__, instance.__class__.__name__))
+    try:
+        if integration.is_pc_sampled_span(span):
+            for idx, prompt in enumerate(prompts):
+                span.set_tag_str("langchain.request.prompts.%d" % idx, integration.trunc(str(prompt)))
+        model = _extract_model_name(instance)
+        if model is not None:
+            span.set_tag_str("langchain.request.model", model)
+        span.set_tag_str("langchain.request.provider", llm_provider)
+        span.set_tag_str("langchain.request.type", "llm")
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
```

Comment:
I am not a big fan of getting tag names dynamically. Can we default to an allowlist for all possible tags. This will be difficult here since `param` and `k` could be anything but maybe we can create a list of supported llm providers?

Reply:
I can't really see the reason for adding an allowlist of llm providers. Can you elaborate?

Reply:
This ties into a suggestion I made earlier of listing all possible LangChain name tags in a constants file and then referencing those values in this integration. This way we can have a clear picture on the data that's captured and sent. 

Say if there is an issue with the following tag: `langchain.request.moonchain_llm.parameters.booo.hoooo` it would be very difficult to find where this tag is set in our library. You will need to step through the whole integration to find where this tag is generated dynamically. It's not the biggest concern but something to think about.

Reply:
Good point - I agree this isn't ideal, but unfortunately with the nature of LangChain (it is an interface for 100+ LLM/embeddings/vectorstores providers) and each different provider has their own unique set of parameter names, meaning it would be pretty infeasible to keep up-to-date with this information by storing them as a constant.

The other alternative is to just stringify the parameters dictionary and store that as one big tag to avoid the cardinality, but that has its own downsides (granularity, visualization). WDYT?